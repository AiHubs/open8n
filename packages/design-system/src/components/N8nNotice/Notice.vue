<template>
	<div :id="id" :class="classes" role="alert">
		<div class="notice-content">
			<n8n-text size="small">
				<slot>
					<span
						v-if="expandFromContent"
						v-html="preExpansionText"
						:class="expanded ? $style['expanded'] : $style['truncated']"
						:id="`${id}-content`"
						role="pre-expansion-text-region"
					/>
					<a
						v-if="expandFromContent"
						v-html="expansionText"
						role="expansion-text-region"
						:aria-controls="`${id}-content`"
						:aria-expanded="canTruncate && !expanded ? 'false' : 'true'"
						@click="toggleExpanded"
					/>
					<span
						:class="expanded ? $style['expanded'] : $style['truncated']"
						:id="`${id}-content`"
						role="region"
						v-html="expandFromContent ? postExpansionText : sanitizedContent"
					/>
					<span v-if="canTruncate">
						<a
							role="button"
							@click="toggleExpanded"
							:aria-controls="`${id}-content`"
							:aria-expanded="canTruncate && !expanded ? 'false' : 'true'"
						>
							{{ expanded ? t('notice.showLess') : expandFromContent ? '' : t('notice.showMore') }}
						</a>
					</span>
				</slot>
			</n8n-text>
		</div>
	</div>
</template>

<script lang="ts">
import Vue from 'vue';
import sanitizeHtml from 'sanitize-html';
import N8nText from "../../components/N8nText";
import Locale from "../../mixins/locale";
import {uid} from "../../utils";

const DEFAULT_TRUNCATION_MAX_LENGTH = 150;

export default Vue.extend({
	name: 'n8n-notice',
	directives: {},
	mixins: [
		Locale,
	],
	props: {
		id: {
			type: String,
			default: () => uid('notice'),
		},
		theme: {
			type: String,
			default: 'warning',
		},
		truncateAt: {
			type: Number,
			default: 150,
		},
		truncate: {
			type: Boolean,
			default: false,
		},
		content: {
			type: String,
			default: '',
		},
		expandFromContent: {
			type: Boolean,
			default: false,
		},
		expansionTextPattern: {
			type: RegExp,
			default: () => /./,
		},
	},
	components: {
		N8nText,
	},
	data() {
		return {
			expanded: false,
		};
	},
	computed: {
		classes(): string[] {
			return [
				'notice',
				this.$style['notice'],
				this.$style[this.theme],
			];
		},
		canTruncate(): boolean {
			return this.truncate && this.content.length > this.truncateAt;
		},
		truncatedContent(): string {
			if (!this.canTruncate || this.expanded) {
				return this.content;
			}

			return this.content.slice(0, this.truncateAt as number) + '...';
		},
		sanitizedContent(): string {
			return sanitizeHtml(this.truncatedContent);
		},
		expansionText(): string {
			const match = this.truncatedContent.match(this.expansionTextPattern);

			if (match.length !== 1) {
				throw new Error('Expansion text must appear once in content');
			}

			return match[0];
		},
		preExpansionText(): string {
			const pre = this.truncatedContent.split(this.expansionTextPattern).shift();

			return sanitizeHtml(pre);
		},
		postExpansionText(): string {
			const post = this.truncatedContent.split(this.expansionTextPattern).pop();

			return sanitizeHtml(post);
		},
	},
	methods: {
		toggleExpanded() {
			this.expanded = !this.expanded;
		},
	},
});
</script>

<style lang="scss" module>
.notice {
	display: flex;
	color: var(--custom-font-black);
	margin: 0;
	padding: var(--spacing-2xs);
	background-color: var(--background-color);
	border-width: 1px 1px 1px 7px;
	border-style: solid;
	border-color: var(--border-color);
	border-radius: var(--border-radius-small);
	line-height: var(--font-line-height-compact);

	a {
		font-weight: var(--font-weight-bold);
	}

	span {
		line-height: var(--font-line-height-compact);
	}
}

.warning {
	--border-color: var(--color-warning-tint-1);
	--background-color: var(--color-warning-tint-2);
}

.danger {
	--border-color: var(--color-danger-tint-1);
	--background-color: var(--color-danger-tint-2);
}

.success {
	--border-color: var(--color-success-tint-1);
	--background-color: var(--color-success-tint-2);
}

.info {
	--border-color: var(--color-info-tint-1);
	--background-color: var(--color-info-tint-2);
}

.expanded {
	+ span {
		margin-top: var(--spacing-4xs);
		display: block;
	}
}

.truncated {
	display: inline;
}
</style>
