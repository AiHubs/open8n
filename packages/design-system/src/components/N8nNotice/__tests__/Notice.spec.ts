import {fireEvent, render} from '@testing-library/vue';
import N8nNotice from "../Notice.vue";

describe('components', () => {
	describe('N8nNotice', () => {
		it('should render correctly', () => {
			const wrapper = render(N8nNotice, {
				props: {
					id: 'notice',
				},
				slots: {
					default: 'This is a notice.',
				},
				stubs: ['n8n-text'],
			});
			expect(wrapper.html()).toMatchSnapshot();
		});

		describe('props', () => {
			describe('content', () => {
				it('should render correctly with content prop', () => {
					const wrapper = render(N8nNotice, {
						props: {
							id: 'notice',
							content: 'This is a notice.',
						},
						stubs: ['n8n-text'],
					});
					expect(wrapper.html()).toMatchSnapshot();
				});

				it('should render HTML', () => {
					const wrapper = render(N8nNotice, {
						props: {
							id: 'notice',
							content: '<strong>Hello world!</strong> This is a notice.',
						},
						stubs: ['n8n-text'],
					});

					expect(wrapper.container.querySelectorAll('strong')).toHaveLength(1);
					expect(wrapper.html()).toMatchSnapshot();
				});

				it('should sanitize rendered HTML', () => {
					const wrapper = render(N8nNotice, {
						props: {
							id: 'notice',
							content: '<script>alert(1);</script> This is a notice.',
						},
						stubs: ['n8n-text'],
					});

					expect(wrapper.container.querySelector('script')).not.toBeTruthy();
					expect(wrapper.html()).toMatchSnapshot();
				});
			});
		});

		describe('truncation', () => {
			it('should truncate content longer than 150 characters - with trailing ellipsis', async () => {
				const wrapper = render(N8nNotice, {
					props: {
						id: 'notice',
						truncate: true,
						trailingEllipsis: true,
						content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
					},
					stubs: ['n8n-text'],
				});

				const post = await wrapper.findByRole('post-expansion-region');

				expect(post).toBeVisible();
				expect(post.textContent!.endsWith('...')).toBeTruthy();
			});

			it('should expand truncated text when clicking \'Show more\'', async () => {
				const wrapper = render(N8nNotice, {
					props: {
						id: 'notice',
						truncate: true,
						trailingEllipsis: true,
						content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
					},
					stubs: ['n8n-text'],
				});

				const button = await wrapper.findByRole('button');
				const post = await wrapper.findByRole('post-expansion-region');

				await fireEvent.click(button);

				expect(button).toHaveTextContent('Show less');
				expect(post.textContent!.endsWith('...')).not.toBeTruthy();
			});
		});

		describe('expansion text', () => {

			describe('should expand from content', () => {

				it('expansion text at start - without trailing ellipsis', async () => {
					const wrapper = render(N8nNotice, {
						props: {
							id: 'notice',
							truncate: true,
							expandFromContent: true,
							expansionTextPattern: /\d+ scopes?/,
							content: '2 scopes available for Google Calendar credentials<br>user.profile<br>user.contacts',
						},
						stubs: ['n8n-text'],
					});

					const pre = await wrapper.findByRole('pre-expansion-region');
					const expansionRegion = await wrapper.findByRole('expansion-region');
					const post = await wrapper.findByRole('post-expansion-region');

					expect(pre).toBeVisible();
					expect(expansionRegion).toBeVisible();
					expect(post).toBeVisible();

					expect(pre).toBeEmptyDOMElement();
					expect(expansionRegion).toHaveTextContent('2 scopes');
					expect(post).toHaveTextContent('available for Google Calendar credentials');
					expect(post.textContent?.endsWith('...')).toBe(false);
				});

				it('expansion text at start - with trailing ellipsis', async () => {
					const content = '2 scopes available for Google Calendar credentials<br>user.profile<br>user.contacts';

					const wrapper = render(N8nNotice, {
						props: {
							id: 'notice',
							truncate: true,
							truncateAt: content.indexOf('<br>'),
							trailingEllipsis: true,
							expandFromContent: true,
							expansionTextPattern: /\d+ scopes?/,
							content,
						},
						stubs: ['n8n-text'],
					});

					const pre = await wrapper.findByRole('pre-expansion-region');
					const expansionRegion = await wrapper.findByRole('expansion-region');
					const post = await wrapper.findByRole('post-expansion-region');

					expect(pre).toBeVisible();
					expect(expansionRegion).toBeVisible();
					expect(post).toBeVisible();

					expect(pre).toBeEmptyDOMElement();
					expect(expansionRegion).toHaveTextContent('2 scopes');
					expect(post).toHaveTextContent('available for Google Calendar credentials');
					expect(post.textContent?.endsWith('...')).toBe(true);
				});

				it('expansion text at middle - without trailing ellipsis', async () => {
					const wrapper = render(N8nNotice, {
						props: {
							id: 'notice',
							truncate: true,
							expandFromContent: true,
							expansionTextPattern: /\d+ scopes?/,
							content: 'Google Books has 1 scope available<br>user.books',
						},
						stubs: ['n8n-text'],
					});

					const pre = await wrapper.findByRole('pre-expansion-region');
					const expansion = await wrapper.findByRole('expansion-region');
					const post = await wrapper.findByRole('post-expansion-region');

					expect(pre).toBeVisible();
					expect(expansion).toBeVisible();
					expect(post).toBeVisible();

					expect(pre).toHaveTextContent('Google Books has');
					expect(expansion).toHaveTextContent('1 scope');
					expect(post).toHaveTextContent('available');
					expect(post.textContent?.endsWith('...')).toBe(false);
				});
			});

			it('expansion text at middle - with trailing ellipsis', async () => {
				const content = 'Google Books has 1 scope available<br>user.books';

				const wrapper = render(N8nNotice, {
					props: {
						id: 'notice',
						truncate: true,
						truncateAt: content.indexOf('<br>'),
						trailingEllipsis: true,
						expandFromContent: true,
						expansionTextPattern: /\d+ scopes?/,
						content,
					},
					stubs: ['n8n-text'],
				});

				const pre = await wrapper.findByRole('pre-expansion-region');
				const expansion = await wrapper.findByRole('expansion-region');
				const post = await wrapper.findByRole('post-expansion-region');

				expect(pre).toBeVisible();
				expect(expansion).toBeVisible();
				expect(post).toBeVisible();

				expect(pre).toHaveTextContent('Google Books has');
				expect(expansion).toHaveTextContent('1 scope');
				expect(post).toHaveTextContent('available');
				expect(post.textContent?.endsWith('...')).toBe(true);
			});
		});

		it('expansion text at end - without trailing ellipsis', async () => {
			const content = 'Google Calendar credentials have 1 scope<br>user.profile';

			const wrapper = render(N8nNotice, {
				props: {
					id: 'notice',
					truncate: true,
					truncateAt: content.indexOf('<br>'),
					expandFromContent: true,
					expansionTextPattern: /\d+ scopes?/,
					content,
				},
				stubs: ['n8n-text'],
			});

			const pre = await wrapper.findByRole('pre-expansion-region');
			const expansionRegion = await wrapper.findByRole('expansion-region');
			const post = await wrapper.findByRole('post-expansion-region');

			expect(pre).toBeVisible();
			expect(expansionRegion).toBeVisible();
			expect(post).toBeVisible();

			expect(pre).toHaveTextContent('Google Calendar credentials have');
			expect(expansionRegion).toHaveTextContent('1 scope');
			expect(post).toBeEmptyDOMElement();
			expect(post.textContent?.endsWith('...')).toBe(false);
		});

		it('expansion text at end - with trailing ellipsis', async () => {
			const content = 'Google Calendar credentials have 1 scope<br>user.profile';

			const wrapper = render(N8nNotice, {
				props: {
					id: 'notice',
					truncate: true,
					truncateAt: content.indexOf('<br>'),
					trailingEllipsis: true,
					expandFromContent: true,
					expansionTextPattern: /\d+ scopes?/,
					content,
				},
				stubs: ['n8n-text'],
			});

			const pre = await wrapper.findByRole('pre-expansion-region');
			const expansionRegion = await wrapper.findByRole('expansion-region');
			const post = await wrapper.findByRole('post-expansion-region');

			expect(pre).toBeVisible();
			expect(expansionRegion).toBeVisible();
			expect(post).toBeVisible();

			expect(pre).toHaveTextContent('Google Calendar credentials have');
			expect(expansionRegion).toHaveTextContent('1 scope');
			expect(post).toHaveTextContent('...');
		});
	});
});
