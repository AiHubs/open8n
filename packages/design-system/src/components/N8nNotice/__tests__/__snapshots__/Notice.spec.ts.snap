// Vitest Snapshot v1

exports[`components > N8nNotice > props > content > should render HTML 1`] = `
"<div id=\\"notice\\" role=\\"alert\\" class=\\"notice _notice_12mks_1 _warning_12mks_20\\">
  <div class=\\"notice-content\\">
    <n8n-text-stub size=\\"small\\" tag=\\"span\\">
      <!---->
      <!----><span id=\\"notice-content\\" role=\\"post-expansion-region\\" class=\\"_truncated_12mks_45\\"><strong>Hello world!</strong> This is a notice.</span>
      <!---->
    </n8n-text-stub>
  </div>
</div>"
`;

exports[`components > N8nNotice > props > content > should render correctly with content prop 1`] = `
"<div id=\\"notice\\" role=\\"alert\\" class=\\"notice _notice_12mks_1 _warning_12mks_20\\">
  <div class=\\"notice-content\\">
    <n8n-text-stub size=\\"small\\" tag=\\"span\\">
      <!---->
      <!----><span id=\\"notice-content\\" role=\\"post-expansion-region\\" class=\\"_truncated_12mks_45\\">This is a notice.</span>
      <!---->
    </n8n-text-stub>
  </div>
</div>"
`;

exports[`components > N8nNotice > props > content > should sanitize rendered HTML 1`] = `
"<div id=\\"notice\\" role=\\"alert\\" class=\\"notice _notice_12mks_1 _warning_12mks_20\\">
  <div class=\\"notice-content\\">
    <n8n-text-stub size=\\"small\\" tag=\\"span\\">
      <!---->
      <!----><span id=\\"notice-content\\" role=\\"post-expansion-region\\" class=\\"_truncated_12mks_45\\"> This is a notice.</span>
      <!---->
    </n8n-text-stub>
  </div>
</div>"
`;

exports[`components > N8nNotice > should render correctly 1`] = `
"<div id=\\"notice\\" role=\\"alert\\" class=\\"notice _notice_12mks_1 _warning_12mks_20\\">
  <div class=\\"notice-content\\">
    <n8n-text-stub size=\\"small\\" tag=\\"span\\">This is a notice.</n8n-text-stub>
  </div>
</div>"
`;
